refactor(step-generation): Update the Thermocycler state shape to allow keeping track of ongoing profiles

SyntaxColoring opened the pull request:
## Overview

This is another step towards EXEC-2009.

On the road to getting step-generation to support concurrent Thermocycler profiles, this lets it represent the fact that a profile is currently in progress (i.e., a profile has been started and has not yet been awaited).

## Test Plan and Hands on Testing

Just existing unit tests.

## Changelog

`ThermocyclerModuleState` is expanded:

* The `blockTargetTemp` property is generalized. It's now `currentBlockActivity`.
  
  Formerly, `blockTargetTemp` could only represent "the block is targeting a temperature" and "the block is deactivated." It can now additionally represent "the block is running a Thermocycler profile."
  
  Nothing sets the state to "the block is running a Thermocycler profile" yet, nor does anything look for it. We're just updating the type to allow it.

  * Not only do we store the fact that a profile is running, but also the steps of the profile, and the Protocol Engine task ID that it was started with. In subsequent PRs, this information will help us when we process `waitForTask` commands and need to update the Thermocycler state to end the profile.

* A new property, `numProfilesStarted`, counts the number of times a profile has been started on the Thermocycler. This is going to be used to generate unique Python variable names, like `task_1 = thermocycler.start_execute_profile(...)`, `task_2 = thermocycler_start_execute_profile(...)`, etc.

## Review requests

`protocol-designer/src/step-forms/types.ts` and `protocol-designer/src/step-forms/selectors/index.ts` have some type definitions and constants that seem exactly duplicated from step-generation. (I marked them with some todo comments.) Any reason to keep them, or can we delete the ones in protocol-designer in favor of the ones in step-generation?

## Risk assessment

Medium.

This should add up to no behavioral difference.

However, I'm a little worried because E2E tests (and no other tests) caught something where some code in protocol-designer needed to get updated to keep up with the `blockTargetTemp`->`currentBlockActivity` rename, and, because of a preexisting `// @ts-expect-error`, the type checker didn't catch it. I did a text search and I don't *think* there's anything else like that lurking around, but it spooked me.

SyntaxColoring committed:
Extend the Thermocycler state type so it can keep track of ongoing profiles.

SyntaxColoring committed:
Increment the profile counter every time a profile is started.

SyntaxColoring commented on app/src/organisms/Desktop/ProtocolVisualization/ModuleContainer/index.tsx:
```diff
@@ -37,14 +37,16 @@ export function ModuleContainer({
+                ? t('temperature', {
+                    temp: currentBlockActivity.blockTargetTemp,
+                  })
                 : t('deactivated')}
```
@jerader Heads up for these updates to protocol visualization. As a first pass, I'm saying that while a Thermocycler profile is ongoing, the block will display as "deactivated." I'm guessing we'll want copy or design updates for this to instead say something like "profile ongoing," right? Let me know if you want tickets or todo comments or something.

ddcc4 replied:
Let's write it as a 3-way case from the start so it's more obvious that this is going to handle `profile`. You don't need the translation yet, of course -- a dummy placeholder string is fine.

SyntaxColoring replied:
0c26d08

SyntaxColoring committed:
Merge branch 'edge' into cma_keep_track_of_ongoing_thermocycler_profiles

See if this fixes failing CI somehow.

ddcc4 commented on protocol-designer/src/step-forms/types.ts:
```diff
@@ -53,14 +53,11 @@ export interface TemperatureModuleState {
+// step-generation instead of duplicatively writing it out here.
+// todo(mm, 2025-12-11): In general, does protocol-designer need its own declarations,
+// for these module types, or can we just directly reuse the types in step-generation?
+export type ThermocyclerModuleState = SG_ThermocyclerModuleState
```
Hm, I don't think it's really idiomatic to create an alias like this (and the `SG_` prefix looks REALLY weird). If you just want to use the `ThermocyclerModuleState` from step-generation directly, I think that's fine.

SyntaxColoring replied:
OK, if we're directly importing `ThermocyclerModuleState` from step-generation, then we ought to do that for all of these types. Mind if I throw that in a separate PR?

[Edit] https://github.com/Opentrons/opentrons/pull/20369

ddcc4 commented on step-generation/src/types.ts:
```diff
@@ -81,9 +81,42 @@ export interface TemperatureModuleState {
+        blockTargetTemp: number
+      }
+    | {
+        type: 'blockDeactivated'
```
Do you prefer `blockDeactivated` to `null`?

Also, you may find that you'll want a name for each of these 3 types, so that you can write:
```
currentBlockActivity: TaskActivity | ConstantTempActivity | Deactivated
or
currentBlockActivity: TaskActivity | ConstantTempActivity | null
```

SyntaxColoring replied:
> Do you prefer `blockDeactivated` to `null`?

I worry that `null` invites confusion over whether it means "deactivated" or "unknown".  `lidOpen` uses `null` for "unknown," for example.

No strong preference, though.

ddcc4 replied:
OK, we can keep `type: 'blockDeactivated'`.

I think it'd still be useful for the 3 cases to have named types though, so that we can refer to them.

SyntaxColoring replied:
269787acca34da0b31108eaa0d434a70d222297d

ddcc4 commented on step-generation/src/types.ts:
```diff
@@ -81,9 +81,42 @@ export interface TemperatureModuleState {
+  lidOpen: boolean | null
+
+  /** Useful for generating unique task IDs every time a new profile is started. */
+  numProfilesStarted: number
```
Hm, I'm slightly sad about having a counter here. Is there state anywhere else from which we can infer the next task number to use? Do you keep track of the set of Tasks that have been created anywhere else?

@jerader Where do we keep track of the numbers when assigning `pythonName`s to labware entities?

SyntaxColoring replied:
I think as long as we want variable names like `thermocycler_1_task_1 = thermocycler_1.start_execute_profile(...)` (as opposed to reusing a single variable like `thermocycler_1_profile_task`), we're going to need this counter *somewhere.* Whether it's in `ThermocyclerModuleState` or `InvariantContext` or something else, I dunno.

ddcc4 replied:
Hm. I'd like to give @jerader a chance to weigh in on this.

(I think for labware, we just query for the count all the labware that ever existed, then add 1 to it when we have to assign a pythonName to a new piece of labware.)

jerader replied:
yeah David's correct with how we get the number for the python name. We go one step further where when you delete a labware, for example, the numbers reconfigure to remove that number. Its happening inside its own reducer where the action is called in the `createContainer` thunk, i think.

I'm curious about how you will handle the case of a step being deleted @SyntaxColoring? Does the state update function then update to change all the previous `numProfilesStarted` values?

SyntaxColoring replied:
Supposing we start with:

1. Transfer
2. Run Thermocycler profile
   * This sees `numProfilesStarted` is 0, so it creates task 1 and sets `numProfilesStarted` to 1
4. Transfer
5. Run Thermocycler profile
   * This sees `numProfilesStarted` is 1, so it creates task 2 and sets `numProfilesStarted` to 2

If the user deletes step 2, the first profile, then we end up with:

1. Transfer
2. Transfer
3. Run Thermocycler profile
   * This sees `numProfilesStarted` is 0, so it creates task 1 and sets `numProfilesStarted` to 1

When the step is removed, the nature of how the timeline recomputes means that all of the profiles will be renumbered automatically, which is what we want. I don't think we need to do anything beyond that, do we?

jerader replied:
oh ya, that is a good point that the timeline recomputes when you delete a step. 

so we increment the numProfilesStarted in the state update fn? And we just get the previous robotState number and +1?

SyntaxColoring replied:
Yep, exactly.

SyntaxColoring committed:
Merge branch 'edge' into cma_keep_track_of_ongoing_thermocycler_profiles

Merging for renames to test files.

SyntaxColoring committed:
Leave placeholders for where we ought to show text like "profile in progress".

SyntaxColoring committed:
Merge branch 'edge' into cma_keep_track_of_ongoing_thermocycler_profiles

SyntaxColoring committed:
Fix merge conflicts.

Some new places cropped up that construct or access ThermocyclerModuleState.

ddcc4 commented on app/src/organisms/Desktop/ProtocolVisualization/ModuleContainer/index.tsx:
```diff
@@ -37,15 +37,19 @@ export function ModuleContainer({
+                  })
+                : currentBlockActivity.type === 'profile'
+                  ? t('deactivated') // todo(mm, 2025-12-12): As a placeholder, this is showing the block as 'deactivated' when there's a profile ongoing.
+                  : t('deactivated')}
```
Heh, I was imagining a literal `switch` statement.

jerader replied:
yeah, we try to avoid nested ternaries, so a switch might be easier to read :)

SyntaxColoring replied:
31d1b144dd97d8a8b4388f17c32cd991e04ebb64

ddcc4 commented on protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepTools/ThermocyclerTools/PriorThermocyclerState.tsx:
```diff
@@ -10,21 +10,39 @@ import type { ThermocyclerModuleState } from '@opentrons/step-generation'
+  switch (currentBlockActivity.type) {
+    // The 'profile' case shouldn't actually be reachable. It would mean this component
+    // is being rendered in a point in the timeline where a profile is ongoing, which
+    // would mean a Thermocycler form is being rendered at a point in yhr yimrlinr
```
Wha?

SyntaxColoring replied:
Oops, finger off by one. Should be "the timeline". 😬

ddcc4 commented on protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepTools/ThermocyclerTools/PriorThermocyclerState.tsx:
```diff
@@ -10,21 +10,39 @@ import type { ThermocyclerModuleState } from '@opentrons/step-generation'
+    // would mean a Thermocycler form is being rendered at a point in yhr yimrlinr
+    // where a profile is ongoing, which would mean a Thermocycler profile has a
+    // Thermocycler step nested within it, which shouldn't be possible.
+    case 'blockDeactivated':
```
Go and throw an `Error` then if you're sure this can't happen.

SyntaxColoring replied:
Are you sure?

If something weird happens that *does* make this happen, an `Error` would hit the nearest React error boundary and could make the app unusable. Wouldn't we much prefer for it to just fall through to some other `case` and merely display wrong copy? Or possibly blank (`""`) copy, instead. And possibly with a `console.error()`.

As far as I've observed, that's been our philosophy for this kind of thing in the frontend in general. Am I wrong about that?

jerader replied:
I prefer a console.error here.

SyntaxColoring replied:
After thinking about this some more, this state actually _is_ possible under normal circumstances. You can make a timeline like this:

1. Thermocycler profile step
   2. Start profile
   3. Thermocycler non-profile step
   4. Wait for profile to complete

This sequence is invalid to *export* because you can't do two things at once with a Thermocycler. The UI *will* mark it as an error, but it still lets you create a timeline like that, you just can't export it.

I'll update this to use "deactivated" as placeholder text with an explanatory comment. Not throwing or logging any errors.

jerader commented on protocol-designer/src/pages/Designer/DeckSetup/DeckSetupDetails.tsx:
```diff
@@ -246,9 +246,26 @@ export function DeckSetupDetails(props: DeckSetupDetailsProps): JSX.Element {
             }
+
+            let blockTargetTemp
+            switch (moduleState.currentBlockActivity.type) {
```
nit: i think it would be handy to break this out into a utility. I think PV will eventually need this.

SyntaxColoring replied:
Could you clarify what you're proposing to break out?

All of the stuff in `getModuleInnerProps()`, for all the modules? Or just this new `blockTargetTemp` switch statement?

jerader replied:
i think just the switch statement. But good point that maybe `getModuleInnerProps()` should be a util instead... hm, i guess you can just leave it as is 😄

SyntaxColoring replied:
Looked into this a little more: this chunk of code *almost* exactly duplicates a `getModuleInnerProps` function in PV, but there are minor differences. Leaving them alone for now.

SyntaxColoring committed:
Merge branch 'edge' into cma_keep_track_of_ongoing_thermocycler_profiles

Resolve conflicts where edge deduplicated some of protocol-designer's types with step-generation:

- protocol-designer/src/step-forms/selectors/index.ts
- protocol-designer/src/step-forms/types.ts

SyntaxColoring committed:
Fix Thermocycler steps sometimes accidentally being created with the block activated.

Some code needed to get updated to account for the new property name. Nothing caught this before because it was obscured by a ts-expect-error.

Also, strangely, this bug only happened for commands other than the very first command in the protocol. The very first command in the protocol is handled specially. That seems wrong, so I've left some todo comments.

SyntaxColoring committed:
Update the createPresavedStepForm tests, and fix more annoying typechecking problems.

SyntaxColoring committed:
Add named types for the three block states.

SyntaxColoring committed:
Fix comment typo.

SyntaxColoring committed:
Use switch statements instead of nested ternaries.

jerader reviewed (approved):
LGTM! The way of incrementing the `numProfilesStarted` makes sense and I think it is cleanest to just handle it in the `moduleState` as opposed to invariant context or something. I also smoke tested to make sure things still work as expected with the `moduleState` change.

SyntaxColoring committed:
Correct the switch case unreachability comment.
